src/crimson/osd/backfill_state.cc

```diff
@@ -0,0 +1,523 @@
+  std::map<pg_shard_t, BackfillInterval>& peer_backfill_info) &&
+{
+  std::for_each(std::begin(pbi_targets), std::end(pbi_targets),
+    [this, &peer_backfill_info] (const auto& bt) {
+      peer_backfill_info.at(bt).pop_front();
+    });
+  last_backfill_started = std::move(new_last_backfill_started);
+}
+
+BackfillState::Enqueuing::loop_advancer_t
+BackfillState::Enqueuing::remove_on_peers(const hobject_t& check)
+{
+  // set `new_last_backfill_started` to `check`
+  loop_advancer_t result { {}, check };
+  for (const auto& bt : ps().get_backfill_targets()) {
+    const auto& pbi = bs().peer_backfill_info.at(bt);
+    if (pbi.begin == check) {
+      result.pbi_targets.insert(bt);
+      const auto& version = pbi.objects.begin()->second;
+      bs().progress_tracker->enqueue_drop(pbi.begin);
+      ls().enqueue_drop(bt, pbi.begin, version);
+    }
+  }
+  logger().debug("{}: BACKFILL removing {} from peers {}",
+                 __func__, check, result.pbi_targets);
+  ceph_assert(!result.pbi_targets.empty());
+  return result;
+}
+
+BackfillState::Enqueuing::loop_advancer_t
+BackfillState::Enqueuing::update_on_peers(const hobject_t& check)
+{
+  loop_advancer_t result { {}, bs().backfill_info.begin };
+  const eversion_t& obj_v = bs().backfill_info.objects.begin()->second;
+
+  for (const auto& bt : ps().get_backfill_targets()) {
+    const auto& pbi = bs().peer_backfill_info.at(bt);
+
+    // Find all check peers that have the wrong version
+    if (check == bs().backfill_info.begin && check == pbi.begin) {
+      if (pbi.objects.begin()->second != obj_v) {
+        bs().progress_tracker->enqueue_push(bs().backfill_info.begin);
+        ls().enqueue_push(bt, bs().backfill_info.begin, obj_v);
+      } else {
+        // it's fine, keep it!
+      }
+      result.pbi_targets.insert(bt);
+    } else {
+      const pg_info_t& pinfo = ps().get_peer_info(bt);
+      // Only include peers that we've caught up to their backfill line
+      // otherwise, they only appear to be missing this object
+      // because their pbi.begin > backfill_info.begin.
+      if (bs().backfill_info.begin > pinfo.last_backfill) {
+        bs().progress_tracker->enqueue_push(bs().backfill_info.begin);
+        ls().enqueue_push(bt, bs().backfill_info.begin, obj_v);
+      }
+    }
+  }
+  return result;
+}
+
+BackfillState::Enqueuing::Enqueuing(my_context ctx)
+  : my_base(ctx)
+{
+  logger().debug("{}", __func__);
+
+  // update our local interval to cope with recent changes
+  bs().backfill_info.begin = bs().last_backfill_started;
+  if (bs().backfill_info.version < ps().get_info().log_tail) {
+    logger().debug("{}: bi is old, rescanning of local backfill_info",
+                   __func__);
+    post_event(RequestPrimaryScanning{});
+    return;
+  } else {
+    maybe_update_range();
+  }
+  trim_backfill_infos();
+
+  while (!bs().backfill_info.empty()) {
+    if (!ls().budget_available()) {
+      post_event(RequestWaiting{});
```

Comment:
Why not make enqueue_push() simply wait until some of the in-progress ones complete.  Seems like that would prevent this statemachine from needing to know about it.

Reply:
IIUC *wait until some of the in-progress ones complete* would be blocking in that case, so some threading, green or OS-level, would be required. With the `budget_available()` I wanted to avoid dependency on `seastar::thread`.

Reply:
Oh, I missed that.  Yeah, that's fine.